### ISSUE TITLE
Headless mode Error

### ISSUE THREAD

**Issue by agrignard** [gama-platform/gama#2608]
**Describe the bug**
After 8 years of working on the visualization of GAMA I am running for the very first time a headless mode (doesn't mean i'm done with visualization as this is an attempt to visualize some of GAMA results in a webrowser).

I might have done something wrong but I don't see what. The simulation is running well in a gui mode.   

**To Reproduce**
1. I am following the doc here https://gama-platform.github.io/wiki/Headless.html
2. My .xml can be found here 
https://github.com/CityScope/CS_CityIO_Backend/blob/master/examples/GAMA/cityIO_Gama/headless/cityIO_Gama.xml

**Expected behavior**
Running and sending data to a server

** Actual Behavior**
`````
sh gama-headless.sh /Users/Arno/Projects/GitHub/CS_CityIO_Backend/examples/GAMA/cityIO_Gama/headless/cityIO_GAMA.xml cityIOOutput
gama-headless.sh
******************************************************************
* GAMA version 1.8.0                                           *
* http://gama-platform.org                                       *
* (c) 2007-2016 UMI 209 UMMISCO IRD/UPMC & Partners              *
******************************************************************
2019-03-20 15:45:43.664 java[66112:26878998] [JRSAppKitAWT markAppIsDaemon]: Process manager already initialized: can't fully enable headless mode.
>GAMA plugin loaded in 3742 ms: 	msi.gama.core_1.7.0.201903200540 [11]
>GAMA plugin loaded in 6 ms: 	ummisco.gama.network_1.7.0.201903200540 [202]
>GAMA plugin loaded in 13 ms: 	msi.gaml.extensions.fipa_1.7.0.201903200540 [18]
>GAMA plugin loaded in 7 ms: 	ummisco.gama.serialize_1.7.0.201903200540 [204]
>GAMA plugin loaded in 3 ms: 	msi.gama.headless_1.7.0.201903200540 [13]
>GAMA plugin loaded in 31 ms: 	simtools.gaml.extensions.traffic_1.7.0.201903200540 [199]
>GAMA plugin loaded in 16 ms: 	irit.gaml.extensions.database_1.7.0.201903200540 [7]
>GAMA plugin loaded in 9 ms: 	ummisco.gaml.extensions.maths_1.7.0.201903200540 [210]
>GAMA plugin loaded in 2 ms: 	ummisco.gama.java2d_1.7.0.201903200540 [201]
>GAMA plugin loaded in 7 ms: 	simtools.gaml.extensions.physics_1.7.0.201903200540 [198]
>GAMA plugin loaded in 12 ms: 	ummisco.gaml.extensions.stats_1.7.0.201903200540 [211]
>GAMA plugin loaded in 0 ms: 	msi.gama.lang.gaml_1.7.0.201903200540 [14]
>GAMA plugin loaded in 10 ms: 	ummisco.gama.opengl_1.7.0.201903200540 [203]
>GAMA plugin loaded in 97 ms: 	msi.gaml.architecture.simplebdi_1.7.0.201903200540 [17]
>GAMA total load time 4038 ms.
Number of cpus used:8 (available: 8)
log4j:WARN No appenders could be found for logger (org.eclipse.xtext.parser.antlr.AbstractInternalAntlrParser).
log4j:WARN Please initialize the log4j system properly.
msi.gama.util.file.http.WebbException: Request failed: 404 Not Found
	at msi.gama.util.file.http.Response.ensureSuccess(Response.java:217)
	at msi.gama.util.file.http.Webb.execute(Webb.java:337)
	at msi.gama.util.file.http.Request.asStream(Request.java:382)
	at msi.gama.common.util.FileUtils.fetchToTempFile(FileUtils.java:396)
	at msi.gama.util.file.GamaFile.fetchFromURL(GamaFile.java:118)
	at msi.gama.util.file.GamaFile.<init>(GamaFile.java:71)
	at msi.gama.util.file.GamaFile.<init>(GamaFile.java:56)
	at msi.gama.util.file.GamaJsonFile.<init>(GamaJsonFile.java:57)
	at gaml.additions.GamlAdditions.lambda$1044(GamlAdditions.java:1103)
	at msi.gaml.types.ParametricFileType.createFile(ParametricFileType.java:142)
	at msi.gaml.types.ParametricFileType.cast(ParametricFileType.java:85)
	at msi.gaml.types.ParametricFileType.cast(ParametricFileType.java:1)
	at msi.gaml.types.ParametricType.cast(ParametricType.java:421)
	at msi.gaml.types.ParametricType.cast(ParametricType.java:1)
	at msi.gaml.operators.Cast.as(Cast.java:168)
	at gaml.additions.GamlAdditions.lambda$967(GamlAdditions.java:1025)
	at msi.gaml.expressions.BinaryOperator._value(BinaryOperator.java:78)
	at msi.gaml.expressions.AbstractExpression.value(AbstractExpression.java:70)
	at msi.gaml.expressions.UnaryOperator._value(UnaryOperator.java:78)
	at msi.gaml.expressions.AbstractExpression.value(AbstractExpression.java:70)
	at msi.gaml.statements.SetStatement.privateExecuteIn(SetStatement.java:139)
	at msi.gaml.statements.AbstractStatement.executeOn(AbstractStatement.java:42)
	at msi.gama.runtime.ExecutionScope.execute(ExecutionScope.java:384)
	at msi.gama.runtime.IScope.execute(IScope.java:505)
	at msi.gaml.statements.AbstractStatementSequence.privateExecuteIn(AbstractStatementSequence.java:55)
	at msi.gaml.statements.TryStatement.privateExecuteIn(TryStatement.java:137)
	at msi.gaml.statements.AbstractStatement.executeOn(AbstractStatement.java:42)
	at msi.gaml.statements.AbstractStatementSequence.executeOn(AbstractStatementSequence.java:44)
	at msi.gama.runtime.ExecutionScope.execute(ExecutionScope.java:384)
	at msi.gama.runtime.IScope.execute(IScope.java:505)
	at msi.gaml.statements.AbstractStatementSequence.privateExecuteIn(AbstractStatementSequence.java:55)
	at msi.gaml.statements.AbstractStatementSequenceWithArgs.privateExecuteIn(AbstractStatementSequenceWithArgs.java:58)
	at msi.gaml.statements.AbstractStatement.executeOn(AbstractStatement.java:42)
	at msi.gaml.statements.AbstractStatementSequence.executeOn(AbstractStatementSequence.java:44)
	at msi.gama.runtime.ExecutionScope.execute(ExecutionScope.java:384)
	at msi.gama.metamodel.agent.MinimalAgent.init(MinimalAgent.java:211)
	at msi.gama.kernel.experiment.ExperimentAgent.init(ExperimentAgent.java:254)
	at msi.gama.runtime.ExecutionScope.lambda$3(ExecutionScope.java:438)
	at msi.gama.runtime.ExecutionScope.runAndCatch(ExecutionScope.java:462)
	at msi.gama.runtime.ExecutionScope.pushRunAndCatch(ExecutionScope.java:452)
	at msi.gama.runtime.ExecutionScope.init(ExecutionScope.java:438)
	at msi.gama.kernel.experiment.ExperimentScheduler.schedule(ExperimentScheduler.java:161)
	at msi.gama.kernel.experiment.ExperimentAgent.schedule(ExperimentAgent.java:306)
	at msi.gama.kernel.experiment.ExperimentPlan.open(ExperimentPlan.java:501)
	at msi.gama.runtime.GAMA.addHeadlessExperiment(GAMA.java:150)
	at msi.gama.headless.core.Experiment.loadCurrentExperiment(Experiment.java:81)
	at msi.gama.headless.core.Experiment.setup(Experiment.java:75)
	at msi.gama.headless.job.ExperimentJob.loadAndBuild(ExperimentJob.java:223)
	at msi.gama.headless.runtime.LocalSimulationRuntime$FakeApplication.run(LocalSimulationRuntime.java:201)
`````
**Screenshots**

**Desktop (please complete the following information):**
 - OS: [e.g. macOS High Sierra]
 - PC Model: [if applicable, can be useful for displays, OpenGL problems)]
 - GAMA version: [e.g. git, 1.7.1, etc.]
 - Java version: [e.g. 1.8]
 - Graphics cards / Display system: [can be useful for displays, OpenGL problems]

**Additional context**

**Comment by AlexisDrogoul** [comment]
Is the access to online resources allowed in headless ?

**Comment by agrignard** [comment]
Good question... 

I would think it's an important (not to say a key) feature if you think about using headless as a way to compute something on a grid/server and reading/writing whatever information from/to a client.

**Comment by hqnghi88** [comment]
Not good enough. I dont see how the security policy would be a key feature for gama. It depends on how your server environment is configured. I can test your model on that github.

**Comment by gnoubi** [comment]
Hi

There is no restriction to access remote file if you use corebuilding file manager. Is your server behind a proxy? I never try in headless mode, but I know, proxies has en effect on eclipse web connection. Maybe it is the same for gama-headless 

Nico

Nicolas Marilleau
UMI 209 UMMISCO IRD/UPMC
Chercheur invité Femto-st

Institut de Recherche pour le Développement 
32 avenue Henri Varagnat
93143 Bondy 
Fixe : 0148027901
Mobile: 0688334906

**Comment by agrignard** [comment]
@hqnghi88 here is the xml file to run cityIO_Gama.gaml in a headless mode if you want to have a look
https://github.com/CityScope/CS_CityIO_Backend/blob/master/examples/GAMA/cityIO_Gama/headless/cityIO_Gama.xml

**Comment by AlexisDrogoul** [comment]
Grouped in #2453 with other headless issues.

**Comment by hqnghi88** [comment]
working model

 /**
* Name: cityIO API for GAMA
* Author: Arnaud Grignard
* Description: Gives the basic read and write feature for cityIO
*  Note: The save to geojson is a feature that is not available in the GAMA 1.8 RC2 (you need a newe version)
*/
model CityIOGAMA



global {

	//geometry shape <- square(1 #km);
 	string cityIOurl <-"https://cityio.media.mit.edu/api/table/virtual_table"; 	
 	string VIRTUAL_LOCAL_DATA <- "./../includes/virtual_table.json";
    map<string, unknown> inputMatrixData;
    map<string, unknown> outputMatrixData;
    map<string, unknown> outputSimulationData;
    map<int,rgb> buildingColors <-[-2::#red, -1::#orange,0::rgb(189,183,107), 1::rgb(189,183,107), 2::rgb(189,183,107),3::rgb(230,230,230), 4::rgb(230,230,230), 5::rgb(230,230,230),6::rgb(40,40,40),7::#cyan,8::#green,9::#gray];
    map<string, unknown> header;
    map<string, unknown> spatial;
    bool updateGrid <- false parameter: "Update Grid:" category: "Grid";
	int gridRrefresh <- 100 min: 1 max:1000 parameter: "Refresh Grid rate (cycle):" category: "Grid";
	bool pushGridToCityIO <- false parameter: "Push Grid to cityIO every refresh cycles" category: "Grid";
	bool pushGridToLocalFile <- false parameter: "Push Grid to a local every refresh cycles" category: "Grid";
	
	int agentRrefresh <- 100 min: 1 max:1000 parameter: "Refresh Agent rate (cycle):" category: "Agent";
	bool pushAgentToCityIO <- false parameter: "Push Agent to cityIO every refresh cycles" category: "Agent";
	bool pushAgentToLocalFile <- false parameter: "Push Agent to a local every refresh cycles" category: "Agent";
	
	int nbCols;
	int nbRows;
	int cellSize;

	init {
	 do initGrid;
	 create people number:10{
	 	shape<-circle(1);
	 }
	}
    // Get a cityIO grid from a url and populate matrixData object
	action initGrid{
		write "coucou";
		try {
			inputMatrixData <- json_file(cityIOurl).contents;
			write "coucouu";
		}
		catch {
			inputMatrixData <- json_file(VIRTUAL_LOCAL_DATA).contents;
			write #current_error + " Impossible to read from cityIO  - Connection to Internet lost or cityIO is offline - inputMatrixData is a local version from cityIO_Kendall.json";
		}
//		 inputMatrixData<-data;
		nbCols<-int(inputMatrixData["header"]["spatial"]["ncols"]);
		 nbRows<-int(inputMatrixData["header"]["spatial"]["nrows"]);
		 cellSize<-int(inputMatrixData["header"]["spatial"]["cellSize"]);
        
		spatial <-inputMatrixData["header"]["spatial"];
		loop i from: 0 to: nbCols-1 {
			loop j from: 0 to: nbRows -1{
				cityMatrix cell <- cityMatrix grid_at { i, j };
				cell.type<-int(inputMatrixData["grid"][j*nbCols+i][0]);
				cell.depth<-int(inputMatrixData["grid"][j*nbCols+i][1]);
			}
        } 
		write "toto" +nbCols;
       write length(cityMatrix where (each.type=-1)); 
	}

	action pushGrid (map<string, unknown> _matrixData){
	  outputMatrixData <- _matrixData;
	  map(outputMatrixData["header"]["owner"])["institute"]<-"Gama Platform";
	  map(outputMatrixData["header"]["owner"])["institute"]<-"Gama Platform";
	  map(outputMatrixData["header"]["owner"])["name"]<-"Arnaud Grignard";
	  map(outputMatrixData["header"]["spatial"])["longitude"]<-105.84;
	  map(outputMatrixData["header"]["spatial"])["latitude"]<-21.02;
	  map(outputMatrixData["header"]["spatial"])["physical_longitude"]<-105.84;
	  map(outputMatrixData["header"]["spatial"])["physical_latitude"]<-21.02;
	  
	  if(pushGridToCityIO){
	  	try{
	  	  save(json_file("https://cityio.media.mit.edu/api/table/update/cityIO_Gama", outputMatrixData));		
	  	}catch{
	  	  write #current_error + " Impossible to write to cityIO - Connection to Internet lost or cityIO is offline";	
	  	} 
	  }
	  if(pushGridToLocalFile){
	  	save(json_file("./../includes/cityIO_Gama.json", outputMatrixData));
	  }
	}
	
	action pushAgent(map<string, unknown> _matrixData){
	  outputMatrixData <- _matrixData;
	  map(outputMatrixData["header"]["owner"])["institute"]<-"Gama Platform";
	  map(outputMatrixData["header"]["owner"])["institute"]<-"Gama Platform";
	  map(outputMatrixData["header"]["owner"])["name"]<-"Arnaud Grignard";
	  map(outputMatrixData["header"]["spatial"])["longitude"]<-105.84;
	  map(outputMatrixData["header"]["spatial"])["latitude"]<-21.02;
	  map(outputMatrixData["header"]["spatial"])["physical_longitude"]<-105.84;
	  map(outputMatrixData["header"]["spatial"])["physical_latitude"]<-21.02;
		list projected_points <- people collect ([each.location]);
		list<map> features<-list_with(length(projected_points), map([]));	
		loop i from: 0 to:length(projected_points)-1{
			list unprojected_point <-point(projected_points[i][0] CRS_transform "EPSG:4326");
			map point_geometry<-['coordinates'::unprojected_point, 'type'::'Point'];
			map point_properties<-['property_1'::2];
			map feature<-["type":: "Feature",'geometry'::point_geometry, 'properties'::point_properties, 'id'::i];
			features[i]<-feature;
			}
		map output_geo<-["type":: "FeatureCollection",'features'::features];	
		outputMatrixData["objects"]<-["points"::output_geo];
		if(pushAgentToCityIO){
	  	try{
	  	  save(json_file("https://cityio.media.mit.edu/api/table/update/cityIO_Gama_Agent", outputMatrixData));		
	  	}catch{
	  	  write #current_error + " Impossible to write to cityIO - Connection to Internet lost or cityIO is offline";	
	  	} 
	  }
	  if(pushAgentToLocalFile){
	  	save(json_file("./../includes/cityIO_Gama_Agent.json", outputMatrixData));
	  }
	}

	reflex updateGrid when: ((cycle mod gridRrefresh) = 0 and updateGrid){
		
		write "tatata";
		do initGrid;
		write "tata";
		if(pushGridToCityIO){
		  do pushGrid(inputMatrixData);	
		}
 	}
 	
 	reflex updateAgent when: ((cycle mod agentRrefresh) = 0){
	  do pushAgent(inputMatrixData);	
 	}
}

grid cityMatrix width:nbCols height:nbRows {
	int size;
	int type;
	int depth;
    aspect base{
	  draw shape color:buildingColors[type] depth:depth;
	  draw string(type) color:#black border:#black at:{location.x,location.y,depth+1};		
	}
}

species people skills:[moving]{
	int attribute1;
	int attribute2;
	reflex move{
		do wander;
	}
	aspect default{
		draw shape color:#blue;
	}
}


experiment Display  type: gui autorun:true{
//	action _init_ {
//   		map<string, unknown> data;
//   		try {
//			data <- json_file(cityIOurl).contents;
//		}
//
//		catch {
//			data <- json_file(VIRTUAL_LOCAL_DATA).contents;
//			write #current_error + "Connection to Internet lost or cityIO is offline - CityMatrix is a local version from cityIO_Kendall.json";
//		}
//   		
//   		
//		create CityIOGAMA_model with: [nbCols::int(data["header"]["spatial"]["ncols"]), nbRows::int(data["header"]["spatial"]["nrows"]),cellSize::int(data["header"]["spatial"]["cellSize"]),inputMatrixData::data]{
//			//shape <-rectangle(nbCols*cellSize, nbRows*cellSize);
//			write " " + nbCols +" " + nbRows + " " + cellSize;
//		}
//	}
	parameter p1 var:cityIOurl <-"https://cityio.media.mit.edu/api/table/virtual_table" ;

	output {	
		display cityMatrixView  type:opengl  background:#black {
			species cityMatrix aspect:base;
			species people aspect: default;
		}
	}
}
